File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/eligibility/gate_ledger.py
from __future__ import annotations

from .types import Rule4GateOutcome, Rule4GateResult


def gate_result(
    gate_id: str,
    outcome: Rule4GateOutcome,
    *,
    evidence_item_ids: list[str] | None = None,
    reason_codes: list[str] | None = None,
    limitation_codes: list[str] | None = None,
) -> Rule4GateResult:
    return {
        "gate_id": gate_id,
        "outcome": outcome,
        "evidence_item_ids": sorted(evidence_item_ids or []),
        "reason_codes": sorted(set(reason_codes or [])),
        "limitation_codes": sorted(set(limitation_codes or [])),
    }


def mandatory_gate_blocks_family(gates: list[Rule4GateResult]) -> bool:
    return any(g["outcome"] != "PASS" for g in gates)


def blocking_gate_codes_from(gates: list[Rule4GateResult]) -> list[str]:
    return sorted(g["gate_id"] for g in gates if g["outcome"] != "PASS")

File: packages/clinical-contracts/src/rule4/eligibility/gateLedger.ts
import type { Rule4GateOutcome, Rule4GateResult } from './types.js';

export function gateResult(
  gateId: string,
  outcome: Rule4GateOutcome,
  opts: {
    evidenceItemIds?: readonly string[];
    reasonCodes?: readonly string[];
    limitationCodes?: readonly string[];
  } = {},
): Rule4GateResult {
  return {
    gateId,
    outcome,
    evidenceItemIds: [...(opts.evidenceItemIds ?? [])].sort(),
    reasonCodes: [...new Set(opts.reasonCodes ?? [])].sort(),
    limitationCodes: [...new Set(opts.limitationCodes ?? [])].sort(),
  };
}

export function mandatoryGateBlocksFamily(gates: readonly Rule4GateResult[]): boolean {
  return gates.some((g) => g.outcome !== 'PASS');
}

export function blockingGateCodesFrom(gates: readonly Rule4GateResult[]): string[] {
  return gates
    .filter((g) => g.outcome !== 'PASS')
    .map((g) => g.gateId)
    .sort();
}
